File: backend/src/schemas/question.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from src.db.models import QuestionType

class QuestionBase(BaseModel):
    text: str
    tags: List[str] = Field(default_factory=list)
    type: QuestionType
    difficulty: Optional[str] = None


class QuestionCreate(QuestionBase):
    pass

class QuestionRead(QuestionBase):
    id: str
    created_at: Optional[datetime] = None

    class Config:
        orm_mode = True

class QuestionSearchRequest(BaseModel):
    query: str
    top_n: Optional[int] = 5
    difficulty: Optional[str] = None
    tags: Optional[List[str]] = None


class QuestionUpdate(BaseModel):
    difficulty: Optional[str] = None
    tags: Optional[List[str]] = None
    type: Optional[QuestionType] = None


class BulkQuestionCreate(BaseModel):
    questions: List[QuestionCreate]


class BulkQuestionResponse(BaseModel):
    message: str
    questions: List[QuestionRead]

class TagRequest(BaseModel):
    tags: List[str]

File: frontend/src/types/question.ts
export type QuestionType = "mcq" | "essay" | "file_upload" | "code" | "numerical" | "true_false" | "short_answer" | "multi_response";

export interface Answer {
  id: string;
  text: string;
  is_correct: boolean; // Required
}

export interface AnswerOption {
  id: string;
  text: string;
  is_correct: boolean;
}

export interface AnswerContainer {
  id?: string;
  explanation?: string;
  options: AnswerOption[];
}

export interface Question {
  id: string; 
  type: QuestionType;
  text: string;
  answer?: AnswerContainer;
  difficulty?: string;
  tags?: string[];
  exam_id?: string;
}

export interface QuestionBase {
  text: string;
  tags: string[];
  type: QuestionType;
  difficulty?: string | null;
}

export interface QuestionCreate extends QuestionBase {}

export interface QuestionRead extends QuestionBase {
  id: string;
  created_at?: string | null; 
}

export interface QuestionSearchRequest {
  query: string;
  top_n?: number;
  difficulty?: string | null;
  tags?: string[] | null;
}

export interface QuestionUpdate {
  difficulty?: string | null;
  tags?: string[] | null;
  type?: QuestionType;
}

export interface BulkQuestionCreate {
  questions: QuestionCreate[];
}

export interface BulkQuestionResponse {
  message: string;
  questions: QuestionRead[];
}

export interface TagRequest {
  tags: string[];
}